### DISCUSSION TITLE
errors are stale inside onChange callback with material ui

### DISCUSSION THREAD

**Discussion by davidstellini** [react-hook-form/react-hook-form#3596] [Q&A]
For some reason it seems that inside the onChange callback, the errors are returned from the last keypress, even though getValue() has the latest value?

To reproduce, type in "k99" into the form:

https://codesandbox.io/s/react-hook-form-validation-attempt2-nv1te?file=/src/index.js

Expected behaviour:
k -> no error
k9 -> error
k99 -> error

current:
k -> no error
k9 -> no error
k99 -> error

Am I doing something wrong or is this a bug?

**Comment by bluebill1049** [comment]
errors is going to update in the next render

**Comment by davidstellini** [comment]
Hmm. I need to know on change because I'm making an api call each change if its valid. Will provide an example with use Effect if this must be on render. had an issue there also.

**Comment by davidstellini** [comment]
Ok - here's an example showing you what happens on the next render. I had to use setState to force the render() to trigger, because otherwise it seems to skip one of every two keypresses. Any ideas?

![image](https://user-images.githubusercontent.com/4324419/101132710-1e42c100-3610-11eb-95bf-a588f4a5182d.png)


Code sandbox: https://codesandbox.io/s/react-hook-form-validation-attempt2-nv1te?file=/src/index.js

Also, your docs on trigger, say: `Manually triggers form validation.`
It also returns a promise, so my expectation is that it tells me whether the field is valid.

I was thinking I'd be able to avoid some boilerplate by just doing:

```
            <TextField
              onChange={(e) => {
                onChange(e);
                trigger("attempt1").then((result) => {
                  if (result) {
                    console.log("valid input, making API call");
                  }
```

But result is always false, even if the input is valid? Why is that?
Here's an example for trigger: https://codesandbox.io/s/react-hook-form-validation-attempt5-5s2vg?file=/src/index.js:1230-1458

**Reply by bluebill1049** [reply]
`onChange` prop from `render` is for sending value to hook form, if you need to trigger validation do at `useEffect`

**Reply by davidstellini** [reply]
Yeah. I'm using useEffect in the linked code sandboxes.

**Reply by davidstellini** [reply]
This part:
```
  const [attempt1, setAttempt1] = useState("");
  const attempt1Errors = { ...formState }?.errors?.attempt1;

  useEffect(() => {
    console.log(
      "Field Data = " +
        attempt1 +
        // "\nerrors.attempt1 = " +
        // { ...errors }?.attempt1 +
        "\nformState.errors.attempt1 = " +
        attempt1Errors +
        "\nmaking api call = " +
        !!(attempt1 && !attempt1Errors)
    );
  }, [attempt1, attempt1Errors]);
```

The problem is I get this. Notice the second one:
```
1. Field Data = k
formState.errors.attempt1 = undefined
making api call = true 
2. Field Data = k9
formState.errors.attempt1 = undefined
making api call = true 
3. Field Data = k9
formState.errors.attempt1 = [object Object]
making api call = false 
```
